src/resty/gcp/request/credentials/accesstoken.lua

```diff
@@ -139,37 +170,71 @@ function AccessToken:new(gcpServiceAccount, opts)
     end
 
     if (accessToken) then
-        self.token = accessToken.access_token
+        self._token = accessToken.access_token
         self.expireTime = ngx.now() + accessToken.expires_in
         self.authMethod = authMethod
+        self.gcpServiceAccount = gcpServiceAccount
     else
         ngx.log(ngx.ERR, "[accesstoken] Unable to get accesstoken")
         error("Failed to authenticate")
         return nil
     end
+
+    self.expireWindow = opts.expireWindow or EXPIRY_WINDOW
     return self
 end
 
 function AccessToken:needsRefresh()
-    return self.expireTime < ngx.now()
+    return self.expireTime < (ngx.now() + self.expireWindow)
 end
 
 function AccessToken:refresh()
     local accessToken
     if (self.authMethod == "SA") then
-        local gcpServiceAccount = os.getenv("GCP_SERVICE_ACCOUNT")
-        accessToken = GetAccessTokenBySA(gcpServiceAccount)
+        accessToken = GetAccessTokenBySA(self.gcpServiceAccount)
     elseif (self.authMethod == "WI") then
         accessToken = GetAccessTokenByWI()
     end
     if (accessToken) then
-        self.token = accessToken.access_token
+        self._token = accessToken.access_token
         self.expireTime = ngx.now() + accessToken.expires_in
         return true
     end
     return false
 end
 
+
+function AccessToken:_auto_refresh()
```

Comment:
It seems there is no need to implement an `_auto_refresh`, just merge it with `refresh` function would be okay I think?